book/src/development/lint_passes.md

```diff
@@ -0,0 +1,125 @@
+# Lint passes
+
+> “A wizard is never late, Frodo Baggins. Nor is he early. He arrives precisely when he means to.”
+> - Gandalf
+
+Before working on the logic of a new lint, there is an important decision
+that every Clippy developers must make: to use
+[`EarlyLintPass`][early_lint_pass] or [`LateLintPass`][late_lint_pass].
+It is an either-or kind of decision. But, like Gandalf the wizard said,
+the point of this decision is really about _when_ we mean for the lint to emit.
+
+So we need to ask ourselves: When do we want this lint to emit
+and what information do we need for the lint to work properly?
```

Comment:
I think the question of "when" is less important and actually a bit confusing. It's true, that `EarlyLintPass`es are run first and can stop compilation early in case of deny, but from Clippy's POV there is little benefit in that. As a new contributor, I would be confused by this. On the other hand, it gives a slight explanation why they are distinguished by the time :)